### PR TITLE
Add animate spriteType

### PR DESCRIPTION
**Pull request by jmif**
- Adds new sprite type `animate` to represent new rendering flow.
- Currently based on segmented sprite but could be changed in the future
- Cache shapes based on bounds in animate sprite to support rendering

### CODE REVIEW

File: app/lib/sprites/SpriteExporter.coffee [hunk 1]
```diff
@@ -24,10 +24,14 @@ class SpriteExporter extends CocoClass
 
   build: ->
     spriteSheetBuilder = new createjs.SpriteSheetBuilder()
+
     if @spriteType is 'segmented'
       @renderSegmentedThangType(spriteSheetBuilder)
+    else if @spriteType is 'animate'
+      throw new Error('Animate sprite type not currently supported in sprite exporter')
```
Comment (jmif) [inline app/lib/sprites/SpriteExporter.coffee]: Searched for all occurrences of spriteType in the client side code base and added explicit handling of animate sprite type.

File: app/lib/surface/AnimateSprite.coffee [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+    else if shapeData.sc?
+      shape.graphics.s shapeData.sc
+
+    shape.graphics.ss shapeData.ss... if shapeData.ss?
+    shape.graphics.de shapeData.de... if shapeData.de?
+    shape.graphics.p shapeData.p if shapeData.p?
+
+    shape.setTransform shapeData.t...
+
+    if shapeData.bounds
```
Comment (jmif) [inline app/lib/surface/AnimateSprite.coffee]: This is the only change in this file, the rest is from SegmentedSprite

File: app/lib/surface/AnimateSprite.coffee [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+
+# Put this on MovieClips
+specialGoToAndStop = (frame) ->
+  if frame is @currentFrame and @childrenCopy
+    @addChild(@childrenCopy...)
+  else
+    @gotoAndStop(frame)
+    @childrenCopy = @children.slice(0)
+
+module.exports = class SegmentedSprite extends createjs.Container
```
Comment (jmif) [inline app/lib/surface/AnimateSprite.coffee]: This is a fork of SegmentedSpite and is intended to represent our next generation rendering flow.  We can modify this as necessary without needing to worry about legacy flows.  For now only change is caching shapes.  It's commented below.

File: app/schemas/models/thang_type.coffee [hunk 4]
```diff
@@ -129,7 +129,7 @@ _.extend ThangTypeSchema.properties,
   scale:
     title: 'Scale'
     type: 'number'
-  spriteType: { enum: ['singular', 'segmented'], title: 'Sprite Type' }
+  spriteType: { enum: [ 'singular', 'segmented', 'animate' ], title: 'Sprite Type' }
```
Comment (jmif) [inline app/schemas/models/thang_type.coffee]: I'm not sold on the name `animate` for the new sprite type - can we think of a better one?

File: app/views/editor/thang/ThangTypeEditView.coffee [hunk 5]
```diff
@@ -363,16 +364,28 @@ module.exports = class ThangTypeEditView extends RootView
...
+
+    worker = new AnimateImporterWorker()
+    worker.addEventListener('message', (event) =>
+      worker.terminate()
+      @hideLoading()
+
+      data = event.data
+
+      if (data.output)
+        @thangType.set('spriteType', 'animate')
```
Comment (jmif) [inline app/views/editor/thang/ThangTypeEditView.coffee]: This is not setting properly

File: app/lib/surface/LayerAdapter.coffee [hunk 6]
```diff
@@ -263,10 +264,10 @@ module.exports = LayerAdapter = class LayerAdapter extends CocoClass
       actionNames = (bundle.actionName for bundle in bundleGrouping)
       args = [thangType, colorConfig, actionNames, builder]
       if thangType.get('raw') or thangType.get('prerenderedSpriteSheetData')
-        if (thangType.get('spriteType') or @defaultSpriteType) is 'segmented'
-          @renderSegmentedThangType(args...)
-        else
-          @renderSingularThangType(args...)
+        switch (thangType.get('spriteType') or @defaultSpriteType)
+          when 'segmented' then @renderSegmentedThangType(args...)
+          when 'animate' then throw new Error('animate sprite type not supported')
```
Comment (AndrewJakubowicz) [inline app/lib/surface/LayerAdapter.coffee]: This code path runs and throws errors everywhere. (CInematics, Level Editor, and play view).

This is not a big  problem as this is not going to be seen by users yet.

While testing I noticed that we don't seem to be able to display multiple instances of the animate sprite type.
To repro. Import some art under the `animate` sprite type.
Change the `kind` property to `misc`.
Then make a new level and click and drop the thang into the level. I can only make 1 appear.

**Comment by jmif** [comment]
Closing as we found an alternative solution to this problem.

**Pull request closed without merging**